src/udev/udev-rules.c

```diff
@@ -2960,50 +2960,69 @@ static int udev_rule_apply_token_to_event(
                         return true;
                 }
 
+                /* Apply formatting to the value. */
                 if (!apply_format_value(event, token, value, sizeof(value), "attribute value"))
                         return true;
 
                 if (EVENT_MODE_DESTRUCTIVE(event)) {
-                        log_event_debug(event, token, "Writing \"%s\" to sysfs attribute \"%s\".", value, buf);
-                        r = write_string_file(buf, value,
-                                              WRITE_STRING_FILE_VERIFY_ON_FAILURE |
-                                              WRITE_STRING_FILE_DISABLE_BUFFER |
-                                              WRITE_STRING_FILE_AVOID_NEWLINE |
-                                              WRITE_STRING_FILE_VERIFY_IGNORE_NEWLINE);
+                        log_event_debug(event, token, "Writing \"%s\" to sysfs attribute \"%s\".", value, resolved);
+                        r = sd_device_set_sysattr_value(dev, resolved, value);
                         if (r < 0)
-                                log_event_error_errno(event, token, r, "Failed to write \"%s\" to sysfs attribute \"%s\", ignoring: %m", value, buf);
+                                log_event_error_errno(event, token, r, "Failed to write \"%s\" to sysfs attribute \"%s\", ignoring: %m", value, resolved);
                         else {
-                                event_cache_written_sysattr(event, buf, value);
+                                event_cache_written_sysattr(event, resolved, value);
                                 log_event_done(event, token);
                         }
                 } else {
-                        log_event_debug(event, token, "Running in test mode, skipping writing \"%s\" to sysfs attribute \"%s\".", value, buf);
+                        log_event_debug(event, token, "Running in test mode, skipping writing \"%s\" to sysfs attribute \"%s\".", value, resolved);
 
-                        r = verify_regular_at(AT_FDCWD, buf, /* follow = */ false);
+                        /* If the path points to a regular file, then we assume it is writable, and cache the
+                         * value to make it shown by OPTIONS="dump" or obtained by later ATTR match token. */
+                        r = fd_verify_regular(fd);
                         if (r < 0 && !ERRNO_IS_NEG_PRIVILEGE(r))
-                                log_event_error_errno(event, token, r, "Failed to verify sysfs attribute \"%s\" is a regular file: %m", buf);
-                        else
-                                event_cache_written_sysattr(event, buf, value);
+                                log_event_error_errno(event, token, r, "Failed to verify sysfs attribute \"%s\" is a regular file: %m", resolved);
+                        else {
+                                if (!streq(last_path_component(resolved), "uevent"))
```

Comment:
i wonder if "uevent" really is the only sysfs attr that is weird like this.

but more importantly: you have the same test at another place. I am pretty sure this test should be moved directly into device_cache_sysattr_value(), and result in an early exit of the function, no? seems safer, and is shorter too